rfc/rfc-0006-child-resource-names.md

```diff
@@ -0,0 +1,43 @@
+# RFC-0006: Child Resource Names
+
+**Authors:** Scott Andrews
+
+**Status:**
+
+**Pull Request URL:** [#1369](https://github.com/projectriff/riff/pull/1369)
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+When a controller creates a new resource, it need to either pick a static name, or let kubernetes generate a name. Static names are predictable, but may already be in use. Generated names cannot be predicated, but are guaranteed to be available.
+
+### Anti-Goals
+This RFC does not prescribe which naming scheme to use for any specific resource.
+
+## Solution
+While there will not be a cut and dry answer for each situation, there are heuristics to guide the decision:
+- if a resource is not consumed directly by a non-controlled resource by name, generate the name (e.g. Pods in a ReplicaSet)
+- if the resource can be queried via a label selector instead of by name, generate the name (e.g. Pods fronted by a Service)
+- if a resource's name surfaces in outward facing way that is part of the public API, consider using a static name (e.g. Services in DNS)
+- if the name needs to be stable between different instances of the same resource, use a static name (e.g. immutable resources that cannot be updated)
+- generate the name
+
+Static names may use templates, but must be deterministic. based only on stable values of the parent resource (e.g. `sprintf('%s-suffix', current.name)`).
+
+If a static name collides with an existing resource, the parent resource should reflect the collision in its conditions. 
+
+The name generally should not be chosen by the creator of the parent resource, as it introduces complexity when reconciling updates for names changes as Kubernetes cannot move, or rename a resource.
+
+### User Impact
+No public interface is changing, as the names of child resources are reflected on the parent resource's status. That will continue to be true.
```

Comment:
Changing controller behavior to make some names deterministic when they are not deterministic today (or vice versa) does impact users. Isn't this the [issue](https://github.com/projectriff/system/issues/216) which triggered this RFC?

Reply:
If/when we decide to change that behavior, maybe. The contract for our resources allows children to be discovered in two ways:
1. the name of is exposed in the parent's status
2. the child can be queried by a label selector

Since a static name is not part of that contract, changing the strategy for naming resources is not a breaking change (nice how that works).

...the notable exception being the core Deployer including .spec.serviceName, which slipped in while I was on holiday. Yes, that was the direct impetus for this RFC, but defining these convention is something we should be doing anyway.